src/main/kotlin/com/avail/files/AvailModuleFile.kt

```diff
@@ -30,131 +30,111 @@
+internal class AvailModuleFile : AbstractAvailTextFile
 {
 	/** The String content of the file. */
-	private lateinit var content: A_String
+	lateinit var content: A_String
+		private set
 
 	override val rawContent: ByteArray get() =
 		content.asNativeString().toByteArray(StandardCharsets.UTF_16BE)
 
-	override fun getSaveContent(): ByteArray =
+	override fun getSaveableContent(): ByteArray =
 		content.asNativeString().toByteArray(StandardCharsets.UTF_8)
 
-	init
+	/**
+	 * Construct an [AvailModuleFile].
+	 *
+	 * @param fileWrapper
+	 *   The [ManagedFileWrapper] that wraps this [AvailFile].
+	 * @param charset
+	 *   The [Charset] of the file.
+	 */
+	constructor(
+		fileWrapper: AbstractFileWrapper,
+		charset: Charset = Charsets.UTF_8): super(charset, fileWrapper)
 	{
-		val sourceBuilder = StringBuilder(4096)
-		var filePosition = 0L
-		val input = ByteBuffer.allocateDirect(4096)
-		val decoder = charset.newDecoder()
-		val output = CharBuffer.allocate(4096)
-//		this.file.read<Any>(
-//			input,
-//			0L,
-//			null,
-		SimpleCompletionHandler<Int>(
+		fileWrapper.reference.readFile(true,
+		{ bytes, _ ->
+			val decoder = charset.newDecoder()
+			decoder.onMalformedInput(CodingErrorAction.REPLACE)
+			decoder.onUnmappableCharacter(CodingErrorAction.REPLACE)
+			try
 			{
-				try
-				{
-					var moreInput = true
-					if (value == -1)
-					{
-						moreInput = false
-					}
-					else
-					{
-						filePosition += value.toLong()
-					}
-					input.flip()
-
-					// TODO not sure if we should care about result
-					val result = decoder.decode(
-						input, output, !moreInput)
-					// If the decoder didn't consume all of the bytes,
-					// then preserve the unconsumed bytes in the next
-					// buffer (for decoding).
-					if (input.hasRemaining())
-					{
-						input.compact()
-					}
-					else
-					{
-						input.clear()
-					}
-					output.flip()
-					sourceBuilder.append(output)
-					// If more input remains, then queue another read.
-					if (moreInput)
-					{
-						output.clear()
-						handler.guardedDo {
-							file.read(input, filePosition, dummy, handler)
-						}
-					}
-					// Otherwise, notify the serverFileWrapper of completion
-					else
-					{
-						decoder.flush(output)
-						sourceBuilder.append(output)
-						content = StringDescriptor.stringWithSurrogatesFrom(
-							sourceBuilder.toString())
-						serverFileWrapper.notifyReady()
-					}
-				}
-				catch (e: IOException)
-				{
-					TODO("Handle AvailServerTextFile read decode bytes fail")
-				}
-			},
+				content = StringDescriptor.stringWithSurrogatesFrom(
+					decoder.decode(ByteBuffer.wrap(bytes)).toString())
+				fileWrapper.notifyReady()
+			}
+			catch (e: Throwable)
 			{
-				TODO("Handle AvailServerTextFile read fail")
-			}).guardedDo { file.read(input, 0L, dummy, handler) }
+				System.err.println(
+					"Attempted to decode bytes from supposed text file " +
+						fileWrapper.reference.uri)
+				System.err.println(e)
```

Comment:
The stack trace needs to be printed. It was code clone issue where I used that line. I thought I fixed all of them to print the stack trace, though clearly I missed one.

As for re-throwing this, this is called after a file is read. That happens on a different thread. There is no upper level to catch it. What it was missing was this line:

```kotlin
fileWrapper.notifyOpenFailure(FileErrorCode.DECODER_FAILURE, e)
```

This will notify all interested parties that the file couldn't be opened because a failure during the decoding of the file.